### DISCUSSION TITLE
How do I assign an IP and tag to a virtual machine using the rest API?

### DISCUSSION THREAD

**Discussion by itinneed2022** [netbox-community/netbox#13906] [Help Wanted!]
Reviewing the swagger page, it seems like the following should work:

```
curl -X 'POST' \
  'https://netbox/api/virtualization/virtual-machines/' \
  -H 'accept: application/json' \
  -H 'Authorization: Token omitted' \
  -H 'Content-Type: application/json' \
  -H 'X-CSRFTOKEN: UJYXwmyJJwGuYhSR1Lzh7TA3f8fzBlLcgaOoBCznbbfrpxxUPx5U1qGFwzmllQkz' \
  -d '{
    "name": "myvm",
    "status": "offline",
    "site": 1,
    "cluster": 1,
    "primary_ip4": 1,
    "vcpus": 2,
    "memory": 8,
    "disk": 60,
    "description": "should work",
    "custom_fields": {
        "OS": "Windows Server 2016"
    }
  }'
```

However all I ever get back is: 
Expecting value: line 3 column 1 (char 2)

What is it missing? primary_ip4 should just be the numeric identifier for the IP address right?

**Comment by atownson** [comment]
This will likely need to be a multi-step process. Here's an untested example based on NetBox v3.5.7:
1. POST the VM without specifying the primary_ip4 - the response includes the new VM ID
```
.../api/virtualization/virtual-machines/
{
    "name": "myvm",
    "status": "offline",
    "site": 1,
    "cluster": 1,
    "vcpus": 2,
    "memory": 8,
    "disk": 60,
    "description": "should work",
    "custom_fields": {
        "OS": "Windows Server 2016"
    }
}
```
2. POST a VMInterface assigned to the VM created - the response includes the new VMInterface ID
```
.../api/virtualization/interfaces/
{
  "virtual_machine": <VM ID>,
  "name": "interface1"
}
```
3. PATCH or POST the IPAddress to associate it the VMInterface
```
.../api/ipam/ip-addresses/
{
    "assigned_object_type": 'virtualization.vminterface',
    "assigned_object_id": <VMInterface ID>
}
```
4. PATCH the VM with the primary_ip4 value
```
.../api/virtualization/virtual-machines/<VM ID>/
{
    "primary_ip4": <IPAddress ID>
{
```

The end result is a VM with an interface, the interface with the IP address associated with it, and the VM lists the IP address as the primary IP address.

**Comment by itinneed2022** [comment]
Following atownson's guide worked! There is simply not a way to assign an IP address when creating a vm, assigning the IP must be iterative.
